File: algorithms/95. Unique Binary Search Trees II/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def generateTrees(self, n):
        """
        :type n: int
        :rtype: List[TreeNode]
        """
        # @StefanPochmann
        def node(val, left, right):
            root = TreeNode(val)
            root.left = left
            root.right = right
            return root
        def gen(start, end):
            return [node(val, left, right)
                    for val in xrange(start, end+1)
                    for left in gen(start, val-1)
                    for right in gen(val+1, end)] or [None]
        return gen(1, n) if n > 0 else []

File: algorithms/95. Unique Binary Search Trees II/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        return n > 0 ? gen(1, n) : vector<TreeNode *>();
    }

    vector<TreeNode *> gen(int start, int end) {
        vector<TreeNode *> result;
        for ( int i = start; i <= end; ++i ) {
            for ( TreeNode * left : gen(start, i - 1) ) {
                for ( TreeNode * right : gen(i + 1, end) ) {
                    TreeNode * root = new TreeNode(i);
                    root->left = left;
                    root->right = right;
                    result.push_back(root);
                }
            }
        }
        return result.size() ? result : vector<TreeNode *> (1, NULL);
    }
};

// 9 / 9 test cases passed.
// Status: Accepted
// Runtime: 16 ms
// beats 41.12 %
